### ISSUE TITLE
Add support for guards

### ISSUE THREAD

**Issue by ipeychev** [metal/metal-router#15]
Let's add support for guards in metal-router. The following guards might be added:
1. CanActivate
2. CanDeactivate
3. Resolve

A guard is a function which returns a Promise or a boolean.
1. If a Promise is returned, the Router should navigate once the Promise is resolved.
2. If the returned value is a boolean, the navigation should happen if the returned value was true.

Why to add these?
1. The developer may easily achieve protected paths - paths, which require authentication. If children are supported, this automatically will protect them too.
2. Deactivating guard helps when a hook for leaving the page should be attached. For example, this might be when there is an unsaved data and the user should be warned before to leave a page.
3. Resolve guard helps fetching data in advance, before navigation to a specific path to happen.
The data might be carried as a payload to the activated route.

Reference: https://angular.io/docs/ts/latest/guide/router.html#!#guards

**Comment by eduardolundgren** [comment]
SGTM. This is possible to be implemented by using Senna `beforeDeactivate` lifecycle.

**Comment by eduardolundgren** [comment]
Hey @ipeychev, I was looking this feature closer and seems very nice and actually not hard to be added, it can hook into Senna screen lifecyle methods for internal implementation.

* CanActivate and CanDeactivate can use https://github.com/liferay/senna.js/blob/master/src/screen/Screen.js#L53.
* Resolve can use https://github.com/liferay/senna.js/blob/master/src/screen/Screen.js#L181.

To simplify usage, those extension points can be exposed on router as options.

**Comment by brunobasto** [comment]
I will look into it.

**Comment by ipeychev** [comment]
Very good, Eduardo, this will be an awesome feature.

**Comment by fernandosouza** [comment]
Hey, guys. This issue is in progress with me. OK? 😃

**Comment by fernandosouza** [comment]
Hey, @eduardolundgren. I'm back on this implementation. I've also notice that the `beforeDeactivate` method should have the following changes.

1 - It should be called in an asynchronous way, to support promises. I suggest right before `doNavigate_`. [implementation](https://github.com/liferay/senna.js/blob/master/src/app/App.js#L366)

2 - We have to change its logic to follow Iliyan`s specifications above. It means that  `beforeDeactivate` should prevent navigation if it returns false, the opposite of the current implementation. For this reason, we have to create a minor version for that.

**Comment by eduardolundgren** [comment]
1- This change is fine. Just note that if it returns a promise the promise must resolve with a boolean value to keep the current contract.
2- The boolean result must behave like it is now, you can customize the contract in metal-router only to implement the suggestion.

**Comment by Robert-Frampton** [comment]
Hey @eduardolundgren @ipeychev , just revisiting our open issues on this project and wanted to clarify some things here.

For `Resolve`, this sounds like what the `fetch`/`fetchUrl` config properties do.

```jsx
<Router
	component={ThirdLevel}
	fetch={true}
	fetchUrl="/some/api.json"
	path='/some/path'
/>
```

This will fetch the`/some/api.json` data and pass it to the component before the component is rendered. Were you hoping for something more generic where you can just pass a function?

**Comment by Robert-Frampton** [comment]
I've sent an initial PR at #29.

`Router` already has a `beforeDeactivateHandler` config property, I modified it so that it can now return a promise.

In addition I added a new `beforeActivateHandler` config property that behaves the same way.

One thing to note, is that both prevent navigation when they resolve to `true` instead of `false`. The reason for this is consistency, since both `beforeDeactivateHandler` and the Senna lifecycle method behave this way already.
